Sources/NIOHTTP2/HTTP2Frame.swift

```diff
@@ -21,108 +21,208 @@ public struct HTTP2Frame {
     /// The payload of this HTTP/2 frame.
     public var payload: FramePayload
 
-    /// The frame flags as an 8-bit integer. To set/unset well-defined flags, consider using the
-    /// other properties on this object (e.g. `endStream`).
-    public var flags: UInt8
-
+    /// The frame flags.
+    public var flags: FrameFlags
+    
     /// The frame stream ID as a 32-bit integer.
     public var streamID: HTTP2StreamID
-
+    
+    private func _hasFlag(_ flag: FrameFlags) -> Bool {
+        return self.flags.contains(flag)
+    }
+    
+    private mutating func _setFlagIfValid(_ flag: FrameFlags) {
+        if self.payload.allowedFlags.contains(flag) {
+            self.flags.formUnion(flag)
+        }
+    }
+    
     // Whether the END_STREAM flag bit is set.
     public var endStream: Bool {
-        get {
-            switch self.payload {
-            case .data, .headers:
-                return (self.flags & UInt8(NGHTTP2_FLAG_END_STREAM.rawValue) != 0)
-            default:
-                return false
-            }
-        }
-        set {
-            switch self.payload {
-            case .data, .headers:
-                self.flags |= UInt8(NGHTTP2_FLAG_END_STREAM.rawValue)
-            default:
-                break
-            }
-        }
+        get { return self._hasFlag(.endStream) }
+        set { self._setFlagIfValid(.endStream) }
     }
-
+    
     // Whether the PADDED flag bit is set.
     public var padded: Bool {
-        get {
-            switch self.payload {
-            case .data, .headers, .pushPromise:
-                return (self.flags & UInt8(NGHTTP2_FLAG_PADDED.rawValue) != 0)
-            default:
-                return false
-            }
-        }
-        set {
-            switch self.payload {
-            case .data, .headers, .pushPromise:
-                self.flags |= UInt8(NGHTTP2_FLAG_PADDED.rawValue)
-            default:
-                break
-            }
-        }
+        get { return self._hasFlag(.padded) }
+        set { self._setFlagIfValid(.padded) }
     }
-
+    
     // Whether the PRIORITY flag bit is set.
     public var priority: Bool {
-        get {
-            if case .headers = self.payload {
-                 return (self.flags & UInt8(NGHTTP2_FLAG_PRIORITY.rawValue) != 0)
-            } else {
-                return false
-            }
-        }
-        set {
-            if case .headers = self.payload {
-                self.flags |= UInt8(NGHTTP2_FLAG_PRIORITY.rawValue)
-            }
-        }
+        get { return self._hasFlag(.priority) }
+        set { self._setFlagIfValid(.priority) }
```

Comment:
While we're here, let's ask the question: now that we're using an `OptionSet` for flags, are these computed properties special enough to stay? @AlanQuatermain, what do you think?

Reply:
There's also the question of what to do if the user wants to set a flag that's invalid for the payload type being used. Silently ignore it? Raise an error? Allow it, but ignore at encode time? Ignore, but raise error when encoding? I've generally gone in the direction of silently ignoring the modification when it's made, in my own implementation.

Reply:
Hm, that doesn't work with `frame.flags.formUnion(someInvalidFlag)` though. It looks like we might have use for an internal `validateFlags()` method or a `validatedFlags` property that would return the intersection of the set flags and those that are actually allowed on the wire. 

Additionally, allowing any flag value to be set means that a client might request padding by setting that flag. I'm in favor of supporting pad bytes when encoding because it's not difficult, but would have it only be used if the client set that flag itself. We wouldn't set it ourselves unless we find a particular use case in testing that would really benefit from it, but really padding seems to be for the benefit of the recipient—our encoding and in-memory representation doesn't cry out for nicely aligned frame sizes. The question, then, is whether the other end does, or whether the intermediary hops would do so. *\*shrug\**

Reply:
Padding is super weird, it's hard to deploy correctly. We should support serialising it when asked for, but let's not give it a special place in the API: I don't want to *encourage* it. ;)

Otherwise, I'm fine with adding a `didSet` that silently strips flags that have no meaning for this frame, or alternatively adding one that traps if you set a bad one.

Reply:
Oh, whoops, I also forgot to say: let's burn the helpers to the ground.

Oh, and actually let's just ignore bad flags, rather than stripping them or trapping. That gives a logical place for extensions to hook in without us having a really rough time, at least early on: they can just set random flag bits if they want to.

Reply:
Yeah, I don't want to encourage use of padding, since it's of limited value to us. That said, if someone wants to go setting that flag, we should honour it.

Reply:
This is a post-hoc comment from me, but I think I disagree with my past self:

> Oh, and actually let's just ignore bad flags, rather than stripping them or trapping. That gives a logical place for extensions to hook in without us having a really rough time, at least early on: they can just set random flag bits if they want to.

I think this was a partially thought-out idea. While it's nice for extension hooking, it leads to a substantially less pleasant programming environment for everyone else. Really, the flags are logically part of the payload of a given frame, and we should treat them as such.

This does raise questions about how we're going to handle extensions, but right now we don't have a good extensibility answer and I don't think it's going to land in the 1.0 release (simply no time to solve that issue). So I think we should consider extensions an explicit goal of the 2.0 release, and in the short term build the best API for non-extensible use-cases we can.

This is mostly here as a historical note and to allow @AlanQuatermain to shout me down if he wants ;).